rstudio/rstudio
# Disable main menu bar when modals are open

### Intent

Addresses #12972 

#### demo
https://github.com/rstudio/rstudio/assets/25834218/fd80ae6b-e5ef-4913-a59d-fb97a9bfb823

### Approach
If a modal is opened, we save the current menu bar and stub the menu bar with a placeholder (it's the same placeholder menu that's used during app startup before the IDE is ready). If any additional (nested or otherwise) modals are opened, we do nothing, since the menu bar is already stubbed. Once all modals are closed, the saved menu bar is restored.

Adds open modal tracking to the app state, so that both GWT and non-GWT modals are considered when restoring the menu bar (only restore the menu bar if all modals are closed).

### Automated Tests
none

### QA Notes

#### Test Scenarios
- GWT modals
    - single modal `Tools > Global Options...`
    - nested modals `Tools > Global Options...` (modal 1), then `Packages > Add...` (modal 2)
- non-GWT modals
    - open choose R or license dialog modals
- electron native modals
    - Check for Updates modal
- combination of GWT, non-GWT and electron native modals open
    - I'm not sure how a user can end up in a state where GWT, non-GWT and native modals are open, but I can provide a custom build to test this if needed (I tested these changes locally with a custom build to check that the modal tracking works)

For each test case, the menu bar should be disabled when the modals are open and should be restored to its pre-modal-opening state after all open modals are closed.

### Documentation
N/A

### Checklist

~- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)~
